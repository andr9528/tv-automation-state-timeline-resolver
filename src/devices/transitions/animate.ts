/**
 * This file contains various classes that handles animations
 * The animators handle arrays of values, so both single magnitudes [opacity] and vectors (coordinates [x, y] or [x, y, z]) are supported
 */

/** Abstract class for Animators */
export abstract class Animator {
	protected positions: number[]

	constructor (startPositions: number[]) {
		this.positions = startPositions
	}
	protected clonePositions () {
		return [...this.positions] // clone
	}
	/** Cause the value to jump to the target value  */
	public jump (target: number[]): number[] {
		if (target.length !== this.positions.length) throw new Error(`Error in Animator.update: target has the wrong length (${target.length}), compared to internal positions (${this.positions.length})`)
		return []
	}
	/** This function is called on every iteraton (frame)
	 * @param target The target value the Animator is aiming towards
	 * @param _timeSinceLastUpdate The delta time since last call to .update()
	 */
	public update (target: number[], _timeSinceLastUpdate: number): number[] {
		if (target.length !== this.positions.length) throw new Error(`Error in Animator.update: target has the wrong length (${target.length}), compared to internal positions (${this.positions.length})`)
		return []
	}
	/** Calculate multi-dimensional distance between values */
	protected getTotalDistanceToTarget (toValue: number[], fromValue: number[]): number {
		return this.hypotenuse(
			fromValue.map((value, index) => {
				const targetPosition = toValue[index]
				return targetPosition - value
			})
		)
	}
	/** Calculate multi-dimensional hypotenuse of a vector */
	protected hypotenuse (vector: number[]): number {
		// Calculate hypotenuse in n dimensions:
		return Math.sqrt(
			vector.map((value) => {
				return Math.pow(value, 2)
			}).reduce((mem, value) => {
				return (mem || 0) + value
			})
		)
	}
}

/** Linear movement towards the target */
export class LinearMovement extends Animator {
	constructor (
		startPositions: number[],
		/** Speed of linear movement [units/ms] */
		private speed: number
	) {
		super(startPositions)
	}
	public jump (target: number[]): number[] {
		super.jump(target)
		this.positions = target
		return this.clonePositions()
	}
	public update (target: number[], timeSinceLastUpdate: number): number[] {
		super.update(target, timeSinceLastUpdate)

		const totalDistanceToTarget = this.getTotalDistanceToTarget(target, this.positions)
		if (totalDistanceToTarget > 0) {
			this.positions.forEach((position, index) => {
				const targetPosition = target[index]
				const distanceToTarget = Math.abs(targetPosition - position)
				const step = (
					this.speed *
					distanceToTarget / totalDistanceToTarget *
					timeSinceLastUpdate
				)
				if (distanceToTarget < step) {
					// The distance left is less than the step, just jump to it then:
					this.positions[index] = targetPosition
				} else {
					// Move towards the target:
					this.positions[index] += step * Math.sign(targetPosition - position)
				}
			})
		}
		return this.clonePositions()
	}
}
/** Simulate physical movement: Accelerate towards target until reaching max speed. Then decelerate in time to stop at target. */
export class PhysicalAcceleration extends Animator {
	private speed: number[]
	private directionChanges: number[]
	constructor (
		/** The starting positions */
		startPositions: number[],
		/** Accelerate towards target with this acceleration. [unit per ] */
		private acceleration: number,
		/** Maximal speed */
		private maxSpeed: number = 2147483648,
		/** Snap to target when distance is less than this value */
		private snapDistance: number = 0
	) {
		super(startPositions)
		this.speed = startPositions.map(() => 0)
		this.directionChanges = startPositions.map(() => 0)
	}
	public jump (target: number[]): number[] {
		super.jump(target)
		this.positions = target
		this.speed = this.speed.map(() => 0)
		return this.clonePositions()
	}
	/**
	 * Update the iteration
	 * @param target Target position(s)
	 * @param timeSinceLastUpdate Time since last update
	 */
	public update (target: number[], timeSinceLastUpdate: number): number[] {
		super.update(target, timeSinceLastUpdate)

		/** Position at next step, given current speed */
		const extrapolatedPositions = this.positions.map((position, index) => position + this.speed[index] * timeSinceLastUpdate * 1)

		const totalDistanceToTarget = this.getTotalDistanceToTarget(target, extrapolatedPositions)

		const totalSpeed = this.hypotenuse(this.speed)

		if (totalDistanceToTarget > 0) {
			let voteToSnap = 0
			extrapolatedPositions.forEach((position, index) => {
				const targetPosition = target[index]

				const distanceToTarget = Math.abs(targetPosition - position)
				/** What direction to accelerate towards */
				const directionToTarget = Math.sign(targetPosition - position)

				// Determine whether to accelerate or decelerate?

				const acceleration = (
					this.acceleration *
					distanceToTarget / totalDistanceToTarget
				)

				const maxSpeed = Math.abs(
					totalSpeed > 0 ?
					(
						this.maxSpeed *
						this.speed[index] / totalSpeed
					) :
					this.maxSpeed
				) || Infinity

				/** Distance to use as threshold for snapping to position */
				// const snapDistance = 4 * acceleration * timeSinceLastUpdate

				/** Absolute value of current speed */
				const speed = Math.abs(this.speed[index])

				/** The time it takes to decelerate to a full stop */
				const timeToStop = speed / acceleration

				/** Mininum distance it takes to decelerate to stop, at the current speed */
				const minimumDistanceToStop = (
					  speed * timeToStop
					- (acceleration * 0.9) * Math.pow(timeToStop, 2) / 2
				)

				const stepAcceleration = directionToTarget * acceleration * timeSinceLastUpdate

				/** Mininum distance it takes to decelerate to stop, at the speed after one step of acceleration */
				const minimumDistanceToStopAfterNextTick = (
					  speed * timeSinceLastUpdate
					+ Math.abs(speed + stepAcceleration) * timeToStop
					- acceleration * Math.pow(timeToStop, 2) / 2
				)

				if (distanceToTarget + Math.max(0, minimumDistanceToStop) <= this.snapDistance) {
					// Vote to snap to target:
					voteToSnap++
					// And decelerate, to prevent wobbling:
					this.speed[index] *= 0.8
				} else {
					if (
						minimumDistanceToStop > distanceToTarget &&
						Math.sign(this.speed[index]) === directionToTarget
					) {
						// Decelerate:
						const speedSign = Math.sign(this.speed[index])
						this.speed[index] -= stepAcceleration
						if (Math.sign(this.speed[index]) !== speedSign) this.speed[index] = 0
					} else if (
						minimumDistanceToStopAfterNextTick < distanceToTarget ||
						Math.sign(this.speed[index]) !== directionToTarget
					) {
						// Accelerate:
						// Apply cap, so that we don't accelerate past the target
						this.speed[index] += this._cap(distanceToTarget / timeSinceLastUpdate, stepAcceleration)
					} else {
						// Neither decelerate or accelerate
					}
				}
				if (
					Math.abs(this.positions[index] - target[index]) <= Math.abs(stepAcceleration) * 8
				) {
					// Apply extra friction when close, to decrease wobbling
					this.speed[index] *= 0.8
				}

				// Cap speed at maxSpeed and apply friction:
				this.speed[index] = this._cap(maxSpeed, this.speed[index])
				this.positions[index] += this.speed[index] * timeSinceLastUpdate

				if (
					this.snapDistance &&
					Math.abs(this.positions[index] - target[index]) <= this.snapDistance &&
					Math.abs(this.speed[index] * timeSinceLastUpdate) <= this.snapDistance * 2
				) {
					this.positions[index] = target[index]
					this.speed[index] = 0
				}

				const newDirectionToTarget = Math.sign(targetPosition - this.positions[index])

				if (Math.sign(directionToTarget) !== Math.sign(newDirectionToTarget)) {
					this.directionChanges[index]++
					// Vote to snap to target:
					// voteToSnap++
				} else {
					this.directionChanges[index] = 0
				}

				if (this.directionChanges[index] > 3) {
					this.positions[index] = target[index]
					this.speed[index] = 0

					this.directionChanges[index] = 0
				}
			})
			if (voteToSnap === this.positions.length) {
				// If everyone wants to snap, lets do that.
				this.positions.forEach((_position, index) => {
					this.positions[index] = target[index]
					this.speed[index] = 0

					this.directionChanges[index] = 0
				})
			}
		}
		return this.clonePositions()
	}
	private _cap (maxValue: number, value: number): number {
		return Math.min(maxValue,
			Math.max(-maxValue,
				value
			)
		)
	}
}
