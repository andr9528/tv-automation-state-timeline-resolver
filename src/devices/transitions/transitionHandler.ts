import { TSRTransitionOptions } from '../../types/src/casparcg'
import * as _ from 'underscore'
import { Animator, LinearMovement, PhysicalAcceleration } from './animate'

interface TransitionHandler {
	values: number[]
	target: number[]
	groups: string[]

	calculatingGroups: {[groupId: string]: {
		animator: Animator
	}}

	activeIterator: NodeJS.Timeout | null
	lastUpdate: number
	updateCallback?: (newValues: number[]) => void
}
export class InternalTransitionHandler {
	private _transitions: {[identifier: string]: TransitionHandler} = {}

	public terminate () {
		// clearInterval(this._interval)
		_.each(this._transitions, (_transition, identifier) => {
			this.clearTransition(identifier)
		})
	}
	public getIdentifiers () {
		return Object.keys(this._transitions)
	}

	public clearTransition (identifier: string) {
		const t = this._transitions[identifier]
		if (t) {
			this._stopTransition(t)
			delete this._transitions[identifier]
		}
	}
	public stopAndSnapTransition (
		identifier: string,
		targetValues: number[]
	) {
		if (!this._transitions[identifier]) {
			this.initTransition(identifier, targetValues)
		}
		const t = this._transitions[identifier]

		this._stopTransition(t)

		t.values = targetValues
	}
	private initTransition (
		identifier: string,
		initialValues: number[]
	) {
		// Set initial values:
		this._transitions[identifier] = {

			values: initialValues,
			target: [], // filled in later
			groups: [], // filled in later

			activeIterator: null,
			lastUpdate: 0,

			calculatingGroups: {}
		}
	}
	public activateTransition (
		identifier: string,
		initialValues: number[],
		targetValues: number[],
		groups: string[],
		options: TSRTransitionOptions,
		animatorTypes: {[groupId: string]:
			{
				type: 'linear' | 'physical',
				options?: TSRTransitionOptions
			}
		},
		updateCallback: (newValues: number[]) => void
	) {

		if (!this._transitions[identifier]) {
			this.initTransition(identifier, initialValues)
		}
		const t = this._transitions[identifier]

		t.updateCallback = updateCallback
		t.groups = groups
		t.target = targetValues

		const getGroupValues = (values: number[], groups: string[], groupId: string) => {
			const vs: number[] = []
			_.each(groups, (g, i) => {
				if (g === groupId) vs.push(values[i])
			})
			return vs
		}
		const setGroupValues = (values: number[], groups: string[], groupId: string, newValues: number[]) => {
			let i2 = 0
			_.each(groups, (g, i) => {
				if (g === groupId) {
					values[i] = newValues[i2]
					i2++
				}
			})
		}

		if (!t.activeIterator) {
			_.each(_.uniq(t.groups), groupId => {
				if (!animatorTypes) animatorTypes = {}
				const animatorType = animatorTypes[groupId + ''] || {}
				const options2 = animatorType.options || options
				t.calculatingGroups[groupId + ''] = {
					animator: (
						animatorType.type === 'physical' ?
						new PhysicalAcceleration(
							getGroupValues(t.values, groups, groupId),
							options2.acceleration || 0.0001,
							options2.maxSpeed || 0.05,
							options2.snapDistance || 1 / 1920
						) :
						new LinearMovement(
							getGroupValues(t.values, groups, groupId),
							options2.linearSpeed || 1 / 1000
						)
					)
				}
			})
			const updateInterval = options.updateInterval || 1000 / 25

			const update = () => {
				let dt = 0
				if (t.lastUpdate) {
					dt = Date.now() - t.lastUpdate
				} else {
					dt = updateInterval
				}
				t.lastUpdate = Date.now()

				let somethingChanged = false
				_.each(_.uniq(t.groups), groupId => {

					const calculatingGroup = t.calculatingGroups[groupId + '']

					const values = getGroupValues(t.values, t.groups, groupId)
					const targetValues = getGroupValues(t.target, t.groups, groupId)
					const newValues = calculatingGroup.animator.update(targetValues, dt)

					if (!_.isEqual(newValues, values)) {
						somethingChanged = true
						setGroupValues(t.values, t.groups, groupId, newValues)
					}
				})
				if (somethingChanged) {
					// Send updateCommand:
					if (t.updateCallback) t.updateCallback(t.values)
				} else {
					// nothing changed
					this._stopTransition(t)
				}
			}

			// Start iterating:
			t.lastUpdate = 0
			t.activeIterator = setInterval(update, updateInterval)
		}
	}
	private _stopTransition (t: TransitionHandler) {
		if (t.activeIterator) {
			clearInterval(t.activeIterator)
			t.activeIterator = null
		}
	}
}
