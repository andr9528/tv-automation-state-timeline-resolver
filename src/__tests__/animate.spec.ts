import * as _ from 'underscore'
import { LinearMovement, PhysicalAcceleration } from '../devices/transitions/animate'

describe('Animate', () => {

	test('LinearMovement', () => {
		{
			// Single
			const animator = new LinearMovement([0], 1)

			let target = [1000]

			const vs: number[][] = []
			const deltaT = 100
			for (let time = 0; time < 1000; time += deltaT) {
				vs.push(animator.update(target, deltaT))
			}
			expect(_.last(vs)).toEqual([1000])
			expect(vs).toMatchSnapshot()

			expect(animator.jump([42])).toEqual([42])
		}
		{
			// x, y
			const animator = new LinearMovement([10, 10], 1)

			let target = [414, 382]

			const vs: number[][] = []
			const deltaT = 25
			for (let i = 0; i < 100; i++) {
				let v = animator.update(target, deltaT)
				vs.push(v)
				if (v[0] === 414) break
			}
			expect(_.last(vs)).toEqual([414, 382])
			expect(vs).toMatchSnapshot()

		}

	})
	test('PhysicalAcceleration', () => {
		{
			// Single
			const animator = new PhysicalAcceleration([0], 0.001, 10000, 1)

			let target = [1000]

			const vs: number[] = []
			const deltaT = 25
			for (let i = 0; i < 300; i++) {
				let v = animator.update(target, deltaT)[0]
				vs.push(v)
				if (v === 1000) break
			}
			expect(vs).toMatchSnapshot()
			expect(_.last(vs)).toEqual(1000)

			expect(animator.jump([42])).toEqual([42])
		}
		{
			// x, y
			const animator = new PhysicalAcceleration([10, 10], 0.001, 1000, 1)

			let target0 = [414, 382]
			let target1 = [-414, 401]

			const vs: number[][] = []
			const deltaT = 25
			for (let i = 0; i < 100; i++) {
				let target = i < 8 ? target0 : target1
				let v = animator.update(target, deltaT)
				vs.push(v)
				if (v[0] === -414 && v[1] === 401) break
			}
			expect(_.last(vs)).toEqual([-414, 401])
			expect(vs).toMatchSnapshot()

		}

	})
})
