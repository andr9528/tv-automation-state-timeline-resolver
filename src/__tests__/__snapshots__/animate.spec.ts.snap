// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Animate LinearMovement 1`] = `
Array [
  Array [
    100,
  ],
  Array [
    200,
  ],
  Array [
    300,
  ],
  Array [
    400,
  ],
  Array [
    500,
  ],
  Array [
    600,
  ],
  Array [
    700,
  ],
  Array [
    800,
  ],
  Array [
    900,
  ],
  Array [
    1000,
  ],
]
`;

exports[`Animate LinearMovement 2`] = `
Array [
  Array [
    28.391015278413054,
    26.934301196954593,
  ],
  Array [
    46.7820305568261,
    43.868602393909185,
  ],
  Array [
    65.17304583523915,
    60.80290359086378,
  ],
  Array [
    83.5640611136522,
    77.73720478781837,
  ],
  Array [
    101.95507639206525,
    94.67150598477296,
  ],
  Array [
    120.3460916704783,
    111.60580718172756,
  ],
  Array [
    138.73710694889135,
    128.54010837868213,
  ],
  Array [
    157.1281222273044,
    145.4744095756367,
  ],
  Array [
    175.51913750571745,
    162.4087107725913,
  ],
  Array [
    193.9101527841305,
    179.3430119695459,
  ],
  Array [
    212.30116806254355,
    196.27731316650048,
  ],
  Array [
    230.6921833409566,
    213.21161436345506,
  ],
  Array [
    249.08319861936965,
    230.14591556040966,
  ],
  Array [
    267.4742138977827,
    247.08021675736427,
  ],
  Array [
    285.8652291761958,
    264.0145179543189,
  ],
  Array [
    304.2562444546088,
    280.9488191512735,
  ],
  Array [
    322.6472597330219,
    297.8831203482281,
  ],
  Array [
    341.03827501143496,
    314.8174215451827,
  ],
  Array [
    359.42929028984804,
    331.7517227421373,
  ],
  Array [
    377.8203055682611,
    348.6860239390919,
  ],
  Array [
    396.21132084667414,
    365.6203251360465,
  ],
  Array [
    414,
    382,
  ],
]
`;

exports[`Animate PhysicalAcceleration 1`] = `
Array [
  0.625,
  1.875,
  3.75,
  6.25,
  9.375,
  13.125,
  17.5,
  22.5,
  28.125,
  34.375,
  41.25,
  48.75,
  56.875,
  65.625,
  75,
  85,
  95.625,
  106.875,
  118.75,
  131.25,
  144.375,
  158.125,
  172.5,
  187.5,
  203.125,
  219.375,
  236.25,
  253.75,
  271.875,
  290.625,
  310,
  330,
  350.625,
  371.875,
  393.75,
  416.25,
  439.375,
  463.125,
  487.5,
  511.25,
  534.375,
  556.875,
  578.75,
  600,
  620.625,
  640.625,
  660,
  678.75,
  696.875,
  714.375,
  731.25,
  748.125,
  764.375,
  780,
  795,
  809.375,
  823.125,
  836.25,
  848.75,
  861.25,
  873.125,
  884.375,
  895,
  905,
  914.375,
  923.125,
  931.875,
  940,
  947.5,
  954.375,
  960.625,
  966.25,
  971.875,
  976.875,
  981.25,
  985,
  988.125,
  991.25,
  993.75,
  995.625,
  996.875,
  998.125,
  998.75,
  1000,
]
`;

exports[`Animate PhysicalAcceleration 2`] = `
Array [
  Array [
    10.459775381960327,
    10.423357529923864,
  ],
  Array [
    11.37932614588098,
    11.270072589771594,
  ],
  Array [
    12.758652291761958,
    12.540145179543188,
  ],
  Array [
    14.597753819603263,
    14.233575299238648,
  ],
  Array [
    16.896630729404894,
    16.35036294885797,
  ],
  Array [
    19.655283021166852,
    18.89050812840116,
  ],
  Array [
    22.873710694889137,
    21.854010837868213,
  ],
  Array [
    26.55191375057175,
    25.24087107725913,
  ],
  Array [
    29.751136777610988,
    29.029232415229856,
  ],
  Array [
    32.46795936336484,
    33.214978723031685,
  ],
  Array [
    34.69898988900345,
    37.79395660933274,
  ],
  Array [
    36.44085920668539,
    42.7619670297879,
  ],
  Array [
    37.69021448811896,
    48.11475650905258,
  ],
  Array [
    38.443713198006975,
    53.848007892395835,
  ],
  Array [
    38.698017148354054,
    59.95733053500635,
  ],
  Array [
    38.44978659150382,
    66.43824982598167,
  ],
  Array [
    37.69567431115659,
    73.28619593046207,
  ],
  Array [
    36.432319671583784,
    80.49649161693532,
  ],
  Array [
    34.65634258589664,
    88.06433901678685,
  ],
  Array [
    32.364337364653274,
    95.98480513893537,
  ],
  Array [
    29.552866406440184,
    104.25280593290026,
  ],
  Array [
    26.218453692527127,
    112.8630886576553,
  ],
  Array [
    22.357578048529238,
    121.81021226954623,
  ],
  Array [
    17.96666613758726,
    131.0885254883453,
  ],
  Array [
    13.042085152432469,
    140.69214213354633,
  ],
  Array [
    7.580135178621194,
    150.6149132398321,
  ],
  Array [
    1.5770412093582795,
    160.85039535677572,
  ],
  Array [
    -4.971055194611845,
    171.3918143073357,
  ],
  Array [
    -12.06810458603649,
    182.23202351470596,
  ],
  Array [
    -19.718158607273793,
    192.7810096470877,
  ],
  Array [
    -27.924700352950715,
    203.045484949187,
  ],
  Array [
    -36.690994464146755,
    213.03192735881535,
  ],
  Array [
    -46.02009480316004,
    222.7465717737305,
  ],
  Array [
    -55.91485140518942,
    232.19540047784082,
  ],
  Array [
    -66.3779165473397,
    241.38413227073218,
  ],
  Array [
    -77.41174974432596,
    250.31820973403393,
  ],
  Array [
    -89.01862143247664,
    259.0027839155055,
  ],
  Array [
    -101.20061503166212,
    267.44269549777636,
  ],
  Array [
    -113.95962696691551,
    275.6424512143236,
  ],
  Array [
    -127.29736406898806,
    283.6061938352805,
  ],
  Array [
    -141.21533752496777,
    291.3376633979536,
  ],
  Array [
    -155.7148521639893,
    298.84014638367006,
  ],
  Array [
    -170.79698924737676,
    306.1164080456324,
  ],
  Array [
    -186.46257992131288,
    313.1686007278678,
  ],
  Array [
    -201.5441764203934,
    319.9981371644357,
  ],
  Array [
    -216.04072031755098,
    326.6078135250845,
  ],
  Array [
    -229.9509979800969,
    333.00088678752303,
  ],
  Array [
    -243.2736140064713,
    339.181171889172,
  ],
  Array [
    -256.0069601548472,
    345.15316390192055,
  ],
  Array [
    -268.14917943628706,
    350.92219279687106,
  ],
  Array [
    -279.6981254744254,
    356.49462093194876,
  ],
  Array [
    -290.65131815346916,
    361.8780968637382,
  ],
  Array [
    -302.20312325765684,
    367.08188371833245,
  ],
  Array [
    -313.1540914171508,
    372.1135660526749,
  ],
  Array [
    -323.5009670543857,
    376.984944848221,
  ],
  Array [
    -333.2399412026255,
    381.7111314274551,
  ],
  Array [
    -342.36665391463555,
    386.31177544933973,
  ],
  Array [
    -350.87635003060046,
    390.8128427660927,
  ],
  Array [
    -358.76440628299383,
    395.2491885252407,
  ],
  Array [
    -366.02770325019117,
    399.66818967250657,
  ],
  Array [
    -372.66778970128,
    404.0399292856058,
  ],
  Array [
    -378.6966690586165,
    408.28108984775594,
  ],
  Array [
    -384.14397470851566,
    412.29334645009413,
  ],
  Array [
    -389.0617709835999,
    415.97357169550486,
  ],
  Array [
    -393.5222939855004,
    419.22773807921703,
  ],
  Array [
    -397.60922533178774,
    421.980851297231,
  ],
  Array [
    -401.4085074029859,
    424.17909255295035,
  ],
  Array [
    -405.00318435003015,
    425.786773275642,
  ],
  Array [
    -408.4725407689439,
    426.7821470775586,
  ],
  Array [
    -411.8940013556143,
    427.1543587904054,
  ],
  Array [
    -415.284506012988,
    426.9023375896364,
  ],
  Array [
    -418.5629446848414,
    426.03544549600497,
  ],
  Array [
    -421.6485026292428,
    424.57406032001046,
  ],
  Array [
    -424.46112689555747,
    422.5504187902358,
  ],
  Array [
    -426.9223880845545,
    420.0098933393553,
  ],
  Array [
    -428.9570189964903,
    417.0126276924172,
  ],
  Array [
    -430.49548285129714,
    413.6353051399769,
  ],
  Array [
    -431.4779335872102,
    410.54341981639436,
  ],
  Array [
    -431.87037724492086,
    407.6577301466726,
  ],
  Array [
    -431.6507401255132,
    404.89846089572194,
  ],
  Array [
    -430.8074334770104,
    402.1799509195226,
  ],
  Array [
    -429.342009369045,
    399.52139814999384,
  ],
  Array [
    -427.2776379640901,
    397.04141507394354,
  ],
  Array [
    -424.67126097447715,
    394.8726472709752,
  ],
  Array [
    -422.50053607653973,
    393.15202258690834,
  ],
  Array [
    -420.674638880912,
    391.952664490798,
  ],
  Array [
    -419.1160724918417,
    391.3182481968809,
  ],
  Array [
    -417.76126086339906,
    391.27468636687314,
  ],
  Array [
    -416.55594174161416,
    391.83798283206545,
  ],
  Array [
    -415.44760387647233,
    393.0187091603677,
  ],
  Array [
    -414,
    394.8236591860252,
  ],
  Array [
    -414,
    396.6286092116828,
  ],
  Array [
    -414,
    397.8085592373403,
  ],
  Array [
    -414,
    398.98850926299787,
  ],
  Array [
    -414,
    399.5434592886554,
  ],
  Array [
    -414,
    401,
  ],
]
`;
